### PR TITLE
update versions sorting to be citus-friendly

### PR DESCRIPTION
**Pull request by katherinebernstein**

### CODE REVIEW

File: src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt [hunk 1]
```diff
@@ -176,7 +176,7 @@ class UpdateDateTimePropertyHash(private val toolbox: Toolbox) : Upgrade {
                 "${updateColumnIfLatestVersion(TEMP_TABLE_NAME, VERSIONS)}," +
                 "${OLD_HASHES_COL.name} = $TEMP_TABLE_NAME.${OLD_HASHES_COL.name} || EXCLUDED.${OLD_HASHES_COL.name}"
 
-        val sortVersions = "ARRAY(SELECT * FROM (SELECT UNNEST(array_agg(DISTINCT ${VERSIONS.name})) as ${VERSIONS.name}) as foo ORDER BY abs(foo.${VERSIONS.name})) as ${VERSIONS.name}"
+        val sortVersions = "ARRAY(SELECT * FROM UNNEST(${VERSIONS.name}) AS foo(${VERSION.name}) ORDER BY abs(foo.${VERSION.name}) AS ${VERSIONS.name}"
```
Comment (you-leee) [inline src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt]: Isn't there the ending `)` missing for ARRAY function?

Comment (katherinebernstein) [reply src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt]: yeah lol this is not what I meant to push up at all...

Comment (you-leee) [reply src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt]: :D

Comment (katherinebernstein) [reply src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt]: undone code has been redone :D

**Review by you-leee — APPROVED** [review]

**Review by dbt — APPROVED** [review]

File: src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt [hunk 2]
```diff
@@ -195,14 +195,14 @@ class UpdateDateTimePropertyHash(private val toolbox: Toolbox) : Upgrade {
...
-                "  ${VALUE_COLUMN.name}, " +
+                "  max(${VALUE_COLUMN.name}), " +
                 "  $newHashComputation AS ${HASH.name}, " +
                 "  array_agg(${HASH.name}) AS ${OLD_HASHES_COL.name} " +
                 "FROM ${DATA.name} " +
                 "WHERE ${PROPERTY_TYPE_ID.name} = ANY(?) " +
                 "AND ${ENTITY_SET_ID.name} = ANY(?) " +
                 "AND length(${HASH.name}) = 16 " +
-                "GROUP BY $keyCols, ${VALUE_COLUMN.name} " +
+                "GROUP BY $keyCols, $newHashComputation " +
```
Comment (dbt) [inline src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt:252]: !dog-chasing-tail.gif

**Review by makosblade — APPROVED** [review]

File: src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt [hunk 3]
```diff
@@ -84,7 +85,12 @@ class UpdateDateTimePropertyHash(private val toolbox: Toolbox) : Upgrade {
         val entitySetIds = toolbox.entitySets.values.filter { entityTypeIds.contains(it.entityTypeId) }.map { it.id }
 
         toolbox.hds.connection.use { conn ->
-            conn.prepareStatement(sql).use { ps ->
+
+//            conn.createStatement().use { stmt ->
```
Comment (makosblade) [inline src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt]: :shame-bell:

Comment (katherinebernstein) [reply src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt]: ding
ding
ding
:shame:

File: src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt [hunk 4]
```diff
@@ -194,14 +218,18 @@ class UpdateDateTimePropertyHash(private val toolbox: Toolbox) : Upgrade {
                 onConflict
     }
 
-    private fun insertRehashedRowsIntoDataTableSql(): String {
+    private fun insertRehashedRowsIntoDataTableSql(partition: Int): String {
         val keyCols = DATA_TABLE_KEY_COLS.joinToString(", ")
         val insertSelectCols = (TEMP_TABLE_UNCHANGED_COLS + HASH).joinToString(", ") { it.name }
 
+        val getBatch = "WITH batch AS ( UPDATE $TEMP_TABLE_NAME set $LAST_MIGRATE = now() WHERE id in (SELECT id from $TEMP_TABLE_NAME WHERE ($LAST_MIGRATE = '-infinity') limit $BATCH_SIZE) RETURNING * )"
```
Comment (dbt) [inline src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt]: i think the inner select should filter on partition.

Comment (katherinebernstein) [reply src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt]: 👍

**Review by dbt — APPROVED** [review]

**Review by pckeyes — APPROVED** [review]

**Review by geekbeast — APPROVED** [review]

**Pull request merged**